### PR TITLE
Senders integration

### PR DESCRIPTION
**Pull request by sstanculeanu**
- added Sender component which creates the internal routine handler with both sender
- made all components from sender package private and only Sender exported
- added managed data handler for heartbeatV2Components
- added randomness on time between sends in order to avoid all peers sending messages in the same time after a restart

### CODE REVIEW

File: factory/heartbeatV2Components_test.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	}()
+
+	args := createMockHeartbeatV2ComponentsFactoryArgs()
+	hcf, err := factory.NewHeartbeatV2ComponentsFactory(args)
+	assert.False(t, check.IfNil(hcf))
+	assert.Nil(t, err)
+
+	hc, err := hcf.Create()
+	assert.NotNil(t, hc)
+	assert.Nil(t, err)
```
Comment (ssd04) [inline factory/heartbeatV2Components_test.go:166]: did you intend to call also `Close` on this test?

File: factory/interface.go [hunk 2]
```diff
@@ -341,6 +341,23 @@ type HeartbeatComponentsHandler interface {
...
+// HeartbeatV2Sender sends heartbeatV2 messages
+type HeartbeatV2Sender interface {
+	Close() error
+	IsInterfaceNil() bool
+}
+
+// HeartbeatV2ComponentsHolder holds the heartbeatV2 components
+type HeartbeatV2ComponentsHolder interface {
+	IsInterfaceNil() bool
+}
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply factory/interface.go:353]: not incomplete, added for consistency only to be used in `node.go`

File: heartbeat/sender/baseSender_test.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/heartbeat/mock"
+	"github.com/stretchr/testify/assert"
+)
+
+func createMockBaseArgs() argBaseSender {
+	return argBaseSender{
+		messenger:                 &mock.MessengerStub{},
+		marshaller:                &mock.MarshallerMock{},
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply heartbeat/sender/baseSender_test.go:14]: `MarshallerMock` is used for both heartbeat v1 and v2..perhaps we can update to the testscommon implementation when cleanup will be done

File: heartbeat/sender/baseSender_test.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+func TestBaseSender_computeRandomDuration(t *testing.T) {
+	t.Parallel()
+
+	bs := createBaseSender(createMockBaseArgs())
+	assert.NotNil(t, bs)
+
+	d1 := bs.computeRandomDuration()
+	d2 := bs.computeRandomDuration()
+	d3 := bs.computeRandomDuration()
+	assert.False(t, d1 == d2)
```
Comment (bogdan-rosianu) [inline heartbeat/sender/baseSender_test.go:31]: this test can fail from time to time as there is a chance that the values are the same

File: heartbeat/sender/heartbeatSender_test.go [hunk 5]
```diff
@@ -40,101 +34,117 @@ func TestNewHeartbeatSender(t *testing.T) {
...
+		sender, err := newHeartbeatSender(args)
 
 		assert.Nil(t, sender)
 		assert.Equal(t, heartbeat.ErrNilCurrentBlockProvider, err)
 	})
+	t.Run("invalid threshold should error", func(t *testing.T) {
+		t.Parallel()
+
+		args := createMockHeartbeatSenderArgs(createMockBaseArgs())
+		args.thresholdBetweenSends = 0
```
Comment (bogdan-rosianu) [inline heartbeat/sender/heartbeatSender_test.go:136]: can you test with 0.001 please?

File: heartbeat/sender/sender.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+			timeBetweenSendsWhenError: args.HeartbeatTimeBetweenSendsWhenError,
+			thresholdBetweenSends:     args.HeartbeatThresholdBetweenSends,
+		},
+		versionNumber:        args.VersionNumber,
+		nodeDisplayName:      args.NodeDisplayName,
+		identity:             args.Identity,
+		peerSubType:          args.PeerSubType,
+		currentBlockProvider: args.CurrentBlockProvider,
+	}
+	err = checkHeartbeatSenderArgs(hbsArgs)
```
Comment (bogdan-rosianu) [inline heartbeat/sender/sender.go:121]: `return checkHeartbeatSenderArgs(hbsArgs)`

File: node/node.go [hunk 7]
```diff
@@ -967,6 +968,11 @@ func (n *Node) GetHeartbeatComponents() mainFactory.HeartbeatComponentsHolder {
 	return n.heartbeatComponents
 }
 
+// GetHeartbeatV2Components returns the heartbeatV2 components
+func (n *Node) GetHeartbeatV2Components() mainFactory.HeartbeatV2ComponentsHolder {
+	return n.heartbeatComponents
```
Comment (bogdan-rosianu) [inline node/node.go:973]: shouldn't you return `heartbeatV2Components` ?

**Review by ssd04 — APPROVED** [review]

**Pull request merged**